inference for algebraic expressions

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63822**

Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

migeed-z committed:
inference for algebraic expressions

[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -1,4 +1,6 @@
 from functools import reduce
+
+import sympy  # type: ignore[import]
```
Wouldn't we also need to guard this import like in the test file?

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -84,6 +86,13 @@ def register(fn):
         return fn
     return register
 
+def register_rule(call_target):
```
The fact that this is named `register_rule` implies it's some sort of superset of the other `register_*_rule` functions. is that the case? Otherwise, should probably name this something else

jfix71 commented on torch/fx/experimental/unify_refinements.py:
```diff
@@ -20,8 +20,16 @@ def infer_symbolic_types(traced):
     to infer all the information such as the case
     for braodcasting.
     """
-    infer_symbolic_types_single_pass(traced)
-    infer_symbolic_types_single_pass(traced)
```
Is `infer_symbolic_types_single_pass` still used, or can it be removed?

migeed-z replied:
I figured it should be there if someone wants to run a single pass? There is a case with broadcasting where doing two passes is needed. I will remove it for now since it would be easy to reconstruct it.

jfix71 commented on test/fx/test_gradual_type.py:
```diff
@@ -934,13 +935,43 @@ def forward(self, x: TensorType((1, 2)), y: TensorType((Dyn, 2))):
+                if norm_layer is None:
+                    norm_layer = torch.nn.BatchNorm2d
+                self.conv1 = conv3x3(inplanes, planes, stride)
+                self.bn1 = norm_layer(planes)
```
This appears to be unused

jfix71 replied:
You marked this resolved but looks like it's still here?

jfix71 reviewed (commented):
Can you add a test where more than one op is chained together that have form some compounded algebraic expression w/ the symbolic shapes for the final output type? Right now you only have single op tests AFAICT.

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations. 

[ghstack-poisoned]

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations. 

[ghstack-poisoned]

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations. 

[ghstack-poisoned]

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations. 

[ghstack-poisoned]

jfix71 commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -9,12 +9,16 @@
-
 from torch.fx.experimental.unification import Var  # type: ignore[attr-defined]
 
+try:
+    import sympy  # type: ignore[import]
+except ImportError:
+    pass
```
Hm, won't this break everything if it can't be imported?

jfix71 reviewed (approved):
LGTM!

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations. 

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518469).

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -9,12 +9,13 @@
-
 from torch.fx.experimental.unification import Var  # type: ignore[attr-defined]
 
+import sympy  # type: ignore[import]
```
How about we do this: we do `try/except` and set `HAS_SYMPY` to True or False based on whether we can import it. Then we can pull the tuple at line 272 out as a constant like `DIMENSION_TYPES =  (int, sympy.Symbol) if HAS_SYMPY else (int,)`. Then we can wrap the definitions of `convert_to_sympy_symbols` and `infer_symbolic_relations` with `if HAS_SYMPY:`

In that case, I think if the user does not have sympy installed, this will degenerate into a still-workable API that just doesn't have the symbolic relations functions in it

jamesr66a reviewed (approved):
Looks good! Let's just fix up the sympy import stuff so it doesn't break things then good to go

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518469).

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518469).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518469).

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518469).

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518469).

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]

migeed-z committed:
Update on "inference for algebraic expressions"


Infer algebraic expressions and add it to our symbolic inferencer. Works for conv2D and can be extended to other operations.

Differential Revision: [D30518469](https://our.internmc.facebook.com/intern/diff/D30518469)

[ghstack-poisoned]
